riscv/triggers.cc

```diff
@@ -153,65 +268,117 @@ module_t::~module_t() {
   }
 }
 
-match_result_t module_t::memory_access_match(action_t * const action, operation_t operation, reg_t address, std::optional<reg_t> data)
+reg_t module_t::tdata1_read(const processor_t * const proc, unsigned index) const noexcept
+{
+  return triggers[index]->tdata1_read(proc);
+}
+
+bool module_t::tdata1_write(processor_t * const proc, unsigned index, const reg_t val) noexcept
+{
+  if (triggers[index]->get_dmode() && !proc->get_state()->debug_mode) {
+    return false;
+  }
+
+  auto xlen = proc->get_xlen();
+
+  // hardware should ignore writes that set dmode to 1 if the previous trigger has both dmode of 0 and chain of 1
+  if (index > 0 && !triggers[index-1]->get_dmode() && triggers[index-1]->get_chain() && get_field(val, CSR_TDATA1_DMODE(xlen)))
+    return false;
+
+  unsigned type = get_field(val, CSR_TDATA1_TYPE(xlen));
+  reg_t tdata1 = val;
+  reg_t tdata2 = triggers[index]->tdata2_read(proc);
+
+  // hardware must zero chain in writes that set dmode to 0 if the next trigger has dmode of 1
+  const bool allow_chain = (index+1 < triggers.size() && triggers[index+1]->get_dmode() && !get_field(val, CSR_TDATA1_DMODE(xlen))) ? false : true;
```

Comment:
What about setting chain=1 on the highest-index trigger? There is no next trigger to chain with. Should that index have chain forced to 0?

Reply:
Should we do this undocumented behavior?

Reply:
When the spec doesn't say, choose the simplest implementation. I wouldn't make a special case for this.